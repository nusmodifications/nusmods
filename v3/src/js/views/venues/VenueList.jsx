// @flow
import React from 'react';
import VenueDetailRow from 'views/venues/VenueDetailRow';

import type { VenueInfo } from 'types/venues';
import type { Venue } from 'types/modules';

import styles from './VenueList.scss';

type Props = {
  venues: VenueInfo,
  expandedVenue: Venue,
  onSelect: (Venue, string, HTMLElement) => void; // Called with venue name and venue URL (/venues/<venue>)
};

// $FlowFixMe: Flow doesn't have Intl typedefs https://github.com/facebook/flow/issues/1270
const collator = new Intl.Collator('en', { sensitivity: 'base', numeric: true });

export default function VenueList(props: Props) {
  const rowRefs: { [Venue]: HTMLElement } = {};
  const { venues, expandedVenue, onSelect } = props;
  const lowercaseExpandedVenue = expandedVenue.toLowerCase();

  // Case-insensitive, natural sort of venue names
  const sortedVenueNames = Object.keys(venues).sort(collator.compare);

  return (
    <ul className={styles.venueList}>
      {sortedVenueNames.map(name => (
        <VenueDetailRow
          key={name}
          name={name}
          availability={venues[name]}
          expanded={name.toLowerCase() === lowercaseExpandedVenue}
          rootElementRef={(row) => {
            if (row) rowRefs[name] = row;
          }}
          onClick={(selectedVenue, venueURL) => onSelect(selectedVenue, venueURL, rowRefs[name])}
        />
      ))}
    </ul>
  );
}
