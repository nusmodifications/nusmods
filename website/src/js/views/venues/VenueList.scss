@import "~styles/utils/modules-entry.scss";

$padding-h: 0.4rem;
$padding-v: 0.5rem;

.heading {
  position: sticky;
  top: 0;
  padding: $padding-v 0 $padding-v $padding-h;
  margin-bottom: 0.5rem;
  border-bottom: 1px solid var(--gray-lighter);
  font-weight: $font-weight-bold;
  font-size: $font-size-lg;
  background: var(--body-bg);

  @include media-breakpoint-down(sm) {
    top: $navbar-height;
  }
}

.venueList {
  list-style: none;
  padding: 0;
}

.subList {
  display: flex;
  flex-wrap: wrap;
  justify-content: stretch;
  list-style: none;
  padding: 0;
  margin-bottom: 0.5rem;

  li {
    flex: 1 1 7rem;
    margin: 1px 1px 0 0;
  }

  :global(.btn) {
    width: 100%;
    padding-right: $padding-h;
    padding-left: $padding-h;
    border-color: transparent;
    text-align: left;
  }
}

@supports (display: grid) {
  .subList {
    display: grid;
    grid-template-columns: repeat(auto-fill, minmax(7rem, 1fr));
    grid-gap: 1px;

    li {
      margin: 0;
    }
  }
}
