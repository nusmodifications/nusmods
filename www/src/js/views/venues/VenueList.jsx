// @flow
import React from 'react';
import classnames from 'classnames';
import { map, groupBy } from 'lodash';
import { Link } from 'react-router-dom';

import type { Venue, VenueDetailList } from 'types/venues';
import { venuePage } from 'views/routes/paths';

import styles from './VenueList.scss';

type Props = {
  venues: VenueDetailList,
  selectedVenue: ?Venue,
};

export default function VenueList(props: Props) {
  const venueListPages = groupBy(props.venues, ([venue]) => venue.charAt(0).toUpperCase());

  return (
    <ul className={styles.venueList}>
      {map(venueListPages, (venues, heading) => (
        <li key={heading}>
          <h3 className={styles.heading}>{heading}</h3>

          <ul className={styles.subList}>
            {venues.map(([venue]) => (
              <li key={venue}>
                <Link
                  to={{
                    pathname: venuePage(venue),
                    search: window.location.search,
                  }}
                  className={classnames(
                    'btn',
                    venue === props.selectedVenue ? 'btn-primary' : 'btn-outline-primary',
                  )}
                >
                  {venue}
                </Link>
              </li>
            ))}
          </ul>
        </li>
      ))}
    </ul>
  );
}
