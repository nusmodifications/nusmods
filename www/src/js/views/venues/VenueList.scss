@import "~styles/utils/modules-entry.scss";

$padding-h: 0.4rem;
$padding-v: 0.5rem;

.heading {
  position: sticky;
  top: 0;
  padding: $padding-v 0 $padding-v $padding-h;
  margin-bottom: 0.5rem;
  font-weight: $font-weight-bold;
  font-size: $font-size-lg;
  border-bottom: 1px solid var(--gray-lighter);
  background: var(--body-bg);

  @include media-breakpoint-down(sm) {
    top: $navbar-height;
  }
}

.venueList {
  padding: 0;
  list-style: none;
}

.subList {
  display: flex;
  flex-wrap: wrap;
  justify-content: stretch;
  padding: 0;
  margin-bottom: 0.5rem;
  list-style: none;

  li {
    flex: 1 1 7rem;
    margin: 1px 1px 0 0;
  }

  :global(.btn) {
    width: 100%;
    padding-right: $padding-h;
    padding-left: $padding-h;
    text-align: left;
    border-color: transparent;
  }
}

@supports (display: grid) {
  .subList {
    display: grid;
    grid-gap: 1px;
    grid-template-columns: repeat(auto-fill, minmax(7rem, 1fr));

    li {
      margin: 0;
    }
  }
}
