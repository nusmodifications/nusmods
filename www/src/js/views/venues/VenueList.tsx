// @flow
import React from 'react';
import classnames from 'classnames';
import { groupBy, toPairs, sortBy } from 'lodash';
import { Link } from 'react-router-dom';

import type { Venue } from 'types/venues';
import { venuePage } from 'views/routes/paths';

import styles from './VenueList.scss';

type Props = {
  venues: Venue[],
  selectedVenue?: ?Venue,
  linkProps?: { [string]: any },
};

export default function VenueList(props: Props) {
  const venueList = groupBy(props.venues, (venue) => venue.charAt(0).toUpperCase());
  const sortedVenueList = sortBy(toPairs(venueList), ([key]) => key);

  return (
    <ul className={styles.venueList}>
      {sortedVenueList.map(([heading, venues]) => (
        <li key={heading}>
          <h3 className={styles.heading}>{heading}</h3>

          <ul className={styles.subList}>
            {venues.map((venue) => (
              <li key={venue}>
                <Link
                  to={{
                    pathname: venuePage(venue),
                    search: window.location.search,
                  }}
                  className={classnames(
                    'btn',
                    venue === props.selectedVenue ? 'btn-primary' : 'btn-outline-primary',
                  )}
                  {...props.linkProps}
                >
                  {venue}
                </Link>
              </li>
            ))}
          </ul>
        </li>
      ))}
    </ul>
  );
}
