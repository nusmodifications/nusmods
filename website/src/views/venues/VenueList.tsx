import { FC, memo, useMemo } from 'react';
import classnames from 'classnames';
import { groupBy, toPairs, sortBy } from 'lodash';
import { Link, LinkProps, useHistory } from 'react-router-dom';

import { Venue } from 'types/venues';
import { venuePage } from 'views/routes/paths';

import styles from './VenueList.scss';

type Props = {
  venues: Venue[];
  selectedVenue?: Venue;
  linkProps?: Omit<LinkProps, 'to'>;
};

const VenueList: FC<Props> = ({ venues, selectedVenue, linkProps }) => {
  const sortedVenueList = useMemo(() => {
    // Added during the horrible COVID-19 times to hide E-Learning venues
    const physicalVenues = venues.filter((venue) => !venue.startsWith('E-Learn'));
    const venueList = groupBy(physicalVenues, (venue) => venue.charAt(0).toUpperCase());
    return sortBy(toPairs(venueList), ([key]) => key);
  }, [venues]);

  const history = useHistory();

  return (
    <ul className={styles.venueList}>
      {sortedVenueList.map(([heading, sortedVenue]) => (
        <li key={heading}>
          <h3 className={styles.heading}>{heading}</h3>
          <ul className={styles.subList}>
            {sortedVenue.map((venue) => (
              <li key={venue}>
                <Link
                  to={{
                    ...history.location,
                    pathname: venuePage(venue),
                  }}
                  className={classnames(
                    'btn',
                    venue === selectedVenue ? 'btn-primary' : 'btn-outline-primary',
                  )}
                  {...linkProps}
                >
                  {venue}
                </Link>
              </li>
            ))}
          </ul>
        </li>
      ))}
    </ul>
  );
};

export default memo(VenueList);
