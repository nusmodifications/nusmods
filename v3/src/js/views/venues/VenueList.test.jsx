// @flow
import React from 'react';
import { mount } from 'enzyme';
import venueInfo from '__mocks__/venueInformation.json';

import type { Venue } from 'types/modules';

import VenueList from './VenueList';

const minProps = {
  venues: venueInfo,
  expandedVenue: '',
  onSelect: () => {},
};

describe('VenueList', () => {
  test('it renders all venues as VenueDetailRows', () => {
    const wrapper = mount(<VenueList {...minProps} />);
    const rows = wrapper.find('VenueDetailRow');
    expect(rows).toHaveLength(Object.keys(venueInfo).length);

    // Case insensitive, natural sort
    const orderedNames = rows.map(r => r.prop('name'));
    expect(orderedNames).toContain('LT1'); // else orderedNames.indexOf('LT1') = -1, which may break test
    expect(orderedNames.indexOf('LT1')).toBeLessThan(orderedNames.indexOf('lt2'));
    expect(orderedNames.indexOf('lt2')).toBeLessThan(orderedNames.indexOf('LT17'));
  });

  test('it expands venues appropriately', () => {
    let wrapper;

    function expectNumberExpandedRows(num: number) {
      return expect(wrapper.find('VenueDetailRow').filterWhere(r => r.prop('expanded'))).toHaveLength(num);
    }

    function expectExpandedVenue(venue: Venue, isExpanded: boolean = true) {
      expect(wrapper.find('VenueDetailRow').filterWhere(r => r.prop('name') === venue).first().prop('expanded')).toBe(isExpanded);
    }

    // Does not expand when there is no expandedVenue
    wrapper = mount(<VenueList {...minProps} />);
    expectNumberExpandedRows(0);

    // Does not expand when no venue names match
    wrapper = mount(<VenueList {...minProps} expandedVenue="covfefe" />);
    expectNumberExpandedRows(0);

    // Expands a valid venue
    wrapper = mount(<VenueList {...minProps} expandedVenue="LT1" />);
    expectNumberExpandedRows(1);
    expectExpandedVenue('LT1');

    // Does not expand partial match
    expectExpandedVenue('LT17', false);

    // Venue name case insensitivity
    wrapper = mount(<VenueList {...minProps} expandedVenue="cqt/SR0622" />);
    expectNumberExpandedRows(1);
    expectExpandedVenue('CQT/SR0622');
  });

  test('it calls onSelect when a row is clicked', () => {
    const mockOnSelect = jest.fn();
    const wrapper = mount(<VenueList {...minProps} onSelect={mockOnSelect} />);
    const lt17Row = wrapper.find('VenueDetailRow').filterWhere(r => r.prop('name') === 'LT17').first();
    lt17Row.find('a').simulate('click');
    expect(mockOnSelect.mock.calls).toHaveLength(1);
    expect(mockOnSelect.mock.calls[0]).toMatchSnapshot();
  });
});
