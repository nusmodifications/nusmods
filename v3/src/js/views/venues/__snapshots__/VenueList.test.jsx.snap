// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VenueList it calls onSelect when a row is clicked 1`] = `
Array [
  "LT17",
  "/venues/LT17",
  <li
    class="venueDetailRow"
  >
    <h4>
      <a
        href="/venues/LT17"
      >
        LT17
      </a>
    </h4>
  </li>,
]
`;
